File: agent/security/run_policy.py
"""Short-lived, signed capability grants for Modal-originated internal calls.

The Next.js internal action routes verify the same compact HMAC format in
``lib/agent/run-policy.ts``.  Python deliberately mints a fresh, narrow grant
at the point of an action instead of carrying a process-wide "agent can act"
token.  The shared internal bearer still authenticates Modal; this token binds
an individual call to a run, workspace, subject, mode, and capability.

This module never reads or logs a secret value.  A missing/misconfigured
``AGENT_RUN_POLICY_SECRET`` returns no grant so shadow rollout compatibility is
preserved; callers that are unattended must still deny local write attempts.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import os
import secrets
import time
import uuid
from collections.abc import Iterable
from typing import Any

RUN_POLICY_HEADER = "x-chippy-run-policy"
RUN_POLICY_AUDIENCE = "chippy-internal-actions"
_MAX_TTL_SECONDS = 900
_DEFAULT_TTL_SECONDS = 300
_VALID_MODES = frozenset({"interactive", "unattended", "voice_control", "sandbox"})
_VALID_CAPABILITIES = frozenset({
    "integration:read",
    "integration:write",
    "team_message:send",
    "task:create_child",
    "task:manage",
    "proposal:decide",
    "sandbox:execute",
})

# Keep this explicit allowlist aligned with
# lib/integrations/action-policy.ts. Unknown provider actions are writes.
_INTEGRATION_READ_ACTIONS = frozenset({
    "GMAIL_FETCH_EMAILS",
    "GMAIL_LIST_THREADS",
    "GMAIL_FETCH_MESSAGE_BY_MESSAGE_ID",
    "OUTLOOK_LIST_MESSAGES",
    "OUTLOOK_GET_MESSAGE",
    "GOOGLECALENDAR_EVENTS_LIST",
    "GOOGLECALENDAR_FIND_FREE_SLOTS",
    "SLACK_LIST_ALL_CHANNELS",
    "SLACK_FETCH_CONVERSATION_HISTORY",
    "SLACK_LIST_ALL_USERS",
    "SLACK_FETCH_MESSAGE_THREAD_FROM_A_CONVERSATION",
    "HUBSPOT_LIST_CONTACTS",
    "HUBSPOT_LIST_DEALS",
    "HUBSPOT_LIST_EMAILS",
    "LINKEDIN_GET_MY_INFO",
    "LINKEDIN_GET_COMPANY_INFO",
    "INSTAGRAM_GET_USER_INFO",
    "INSTAGRAM_LIST_ALL_CONVERSATIONS",
    "FACEBOOK_GET_PAGE_CONVERSATIONS",
    "TWITTER_USER_LOOKUP_ME",
    "NOTION_SEARCH_NOTION_PAGE",
    "NOTION_QUERY_DATABASE",
    "GOOGLESHEETS_VALUES_GET",
    "GOOGLESHEETS_GET_SPREADSHEET_INFO",
    "GOOGLESHEETS_BATCH_GET",
})


def integration_action_capability(slug: str) -> str:
    """Return the one capability required by a reviewed integration action."""
    return "integration:read" if slug in _INTEGRATION_READ_ACTIONS else "integration:write"


def is_unattended_write(mode: str, capability: str) -> bool:
    """Unattended execution proposes writes; it never calls them directly."""
    return mode == "unattended" and capability in {"integration:write", "team_message:send"}


def issue_run_policy(
    *,
    run_id: str,
    space_id: str,
    subject: str,
    mode: str,
    capabilities: Iterable[str],
    ttl_seconds: int = _DEFAULT_TTL_SECONDS,
) -> str | None:
    """Issue a TS-compatible signed grant, or ``None`` when safely unavailable.

    ``None`` intentionally does not raise: AGENT_RUN_POLICY_MODE remains shadow
    during caller migration.  It is not a reason for an unattended caller to
    perform a write; those callers are denied before making an HTTP request.
    """
    secret = os.environ.get("AGENT_RUN_POLICY_SECRET", "")
    normalized_capabilities = list(dict.fromkeys(capabilities))
    if (
        len(secret) < 32
        or not space_id
        or not subject
        or mode not in _VALID_MODES
        or not normalized_capabilities
        or len(normalized_capabilities) > 16
        or any(capability not in _VALID_CAPABILITIES for capability in normalized_capabilities)
    ):
        return None
    try:
        uuid.UUID(run_id)
    except (ValueError, TypeError, AttributeError):
        return None

    now = int(time.time())
    ttl = max(30, min(int(ttl_seconds), _MAX_TTL_SECONDS))
    claims: dict[str, Any] = {
        "v": 1,
        "iss": "chippy",
        "aud": RUN_POLICY_AUDIENCE,
        "runId": run_id,
        "spaceId": space_id,
        "subject": subject,
        "mode": mode,
        "capabilities": normalized_capabilities,
        "depth": 0,
        "iat": now,
        "exp": now + ttl,
        "nonce": secrets.token_urlsafe(24),
    }
    payload = _b64url(json.dumps(claims, separators=(",", ":")).encode("utf-8"))
    signature = hmac.new(secret.encode("utf-8"), payload.encode("ascii"), hashlib.sha256).digest()
    return f"{payload}.{_b64url(signature)}"


def action_headers(context: Any, capability: str) -> dict[str, str]:
    """Bearer-like header map for the exact capability needed by one call.

    ``AgentContext`` is intentionally duck-typed to keep the signing seam
    independent from the legacy Agents SDK.  The only state consumed is the
    run/space/user identity the runtime injected, never tool-model arguments.
    """
    policy_run_id = str(getattr(context, "run_policy_run_id", ""))
    legacy_run_id = str(getattr(context, "run_id", ""))
    run_id = policy_run_id if _is_uuid(policy_run_id) else legacy_run_id
    token = issue_run_policy(
        run_id=run_id,
        space_id=str(getattr(context, "space_id", "")),
        subject=str(getattr(context, "user_id", "")),
        mode=str(getattr(context, "run_mode", "interactive")),
        capabilities=(capability,),
    )
    return {RUN_POLICY_HEADER: token} if token else {}


def _b64url(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _is_uuid(value: str) -> bool:
    try:
        uuid.UUID(value)
    except (ValueError, TypeError, AttributeError):
        return False
    return True

File: lib/agent/run-policy.ts
import 'server-only';

import { createHmac, timingSafeEqual } from 'node:crypto';
import { z } from 'zod';

export const RUN_POLICY_HEADER = 'x-chippy-run-policy';
export const RUN_POLICY_AUDIENCE = 'chippy-internal-actions';

export const RUN_MODES = ['interactive', 'unattended', 'voice_control', 'sandbox'] as const;
export type RunMode = (typeof RUN_MODES)[number];

export const RUN_CAPABILITIES = [
  'integration:read',
  'integration:write',
  'team_message:send',
  'task:create_child',
  'task:manage',
  'proposal:decide',
  'sandbox:execute',
] as const;
export type RunCapability = (typeof RUN_CAPABILITIES)[number];

const claimsSchema = z.object({
  v: z.literal(1),
  iss: z.literal('chippy'),
  aud: z.literal(RUN_POLICY_AUDIENCE),
  runId: z.string().uuid(),
  spaceId: z.string().min(1),
  subject: z.string().min(1),
  mode: z.enum(RUN_MODES),
  capabilities: z.array(z.enum(RUN_CAPABILITIES)).max(16),
  parentRunId: z.string().uuid().optional(),
  depth: z.number().int().min(0).max(8).default(0),
  iat: z.number().int().positive(),
  exp: z.number().int().positive(),
  nonce: z.string().min(16).max(160),
});

export type RunPolicyClaims = z.infer<typeof claimsSchema>;

export type RunPolicyVerification =
  | { ok: true; claims: RunPolicyClaims }
  | { ok: false; reason: 'missing' | 'misconfigured' | 'invalid' | 'expired' | 'scope_mismatch' };

export type RunPolicyEnforcementMode = 'shadow' | 'enforce';

/**
 * Backward-compatible rollout control. Existing customer flows remain in
 * shadow mode until every trusted caller carries a signed grant. Production
 * enablement must explicitly set `AGENT_RUN_POLICY_MODE=enforce`.
 */
export function runPolicyEnforcementMode(): RunPolicyEnforcementMode {
  const value = process.env.AGENT_RUN_POLICY_MODE;
  if (!value || value === 'shadow') return 'shadow';
  if (value === 'enforce') return 'enforce';
  throw new Error(`Invalid AGENT_RUN_POLICY_MODE: ${value}`);
}

function policySecret(): string | null {
  const secret = process.env.AGENT_RUN_POLICY_SECRET;
  return secret && secret.length >= 32 ? secret : null;
}

function signPayload(payload: string, secret: string): Buffer {
  return createHmac('sha256', secret).update(payload).digest();
}

export function signRunPolicy(
  input: Omit<RunPolicyClaims, 'v' | 'iss' | 'aud' | 'iat' | 'exp'> & { ttlSeconds?: number },
): string {
  const secret = policySecret();
  if (!secret) throw new Error('AGENT_RUN_POLICY_SECRET is not configured');

  const now = Math.floor(Date.now() / 1000);
  const claims = claimsSchema.parse({
    ...input,
    v: 1,
    iss: 'chippy',
    aud: RUN_POLICY_AUDIENCE,
    iat: now,
    exp: now + Math.max(30, Math.min(input.ttlSeconds ?? 300, 900)),
  });
  const payload = Buffer.from(JSON.stringify(claims)).toString('base64url');
  return `${payload}.${signPayload(payload, secret).toString('base64url')}`;
}

export function verifyRunPolicy(
  token: string | null,
  expected: { spaceId: string; requiredCapability?: RunCapability },
): RunPolicyVerification {
  if (!token) return { ok: false, reason: 'missing' };
  const secret = policySecret();
  if (!secret) return { ok: false, reason: 'misconfigured' };

  const [payload, signature, extra] = token.split('.');
  if (!payload || !signature || extra) return { ok: false, reason: 'invalid' };

  let provided: Buffer;
  try {
    provided = Buffer.from(signature, 'base64url');
  } catch {
    return { ok: false, reason: 'invalid' };
  }
  const expectedSignature = signPayload(payload, secret);
  if (provided.length !== expectedSignature.length || !timingSafeEqual(provided, expectedSignature)) {
    return { ok: false, reason: 'invalid' };
  }

  let claims: RunPolicyClaims;
  try {
    claims = claimsSchema.parse(JSON.parse(Buffer.from(payload, 'base64url').toString('utf8')));
  } catch {
    return { ok: false, reason: 'invalid' };
  }

  if (claims.exp <= Math.floor(Date.now() / 1000)) return { ok: false, reason: 'expired' };
  if (claims.spaceId !== expected.spaceId) return { ok: false, reason: 'scope_mismatch' };
  if (
    expected.requiredCapability &&
    !claims.capabilities.includes(expected.requiredCapability)
  ) {
    return { ok: false, reason: 'scope_mismatch' };
  }
  return { ok: true, claims };
}

/**
 * An unattended run is never allowed to execute an integration action. It can
 * read from an explicit allowlist and must emit an AgentActionProposal for
 * everything else. Interactive/voice runs still need an explicit capability.
 */
export function mayExecuteIntegrationAction(
  claims: RunPolicyClaims,
  actionClass: 'read' | 'write',
): boolean {
  if (actionClass === 'read') return claims.capabilities.includes('integration:read');
  return (
    (claims.mode === 'interactive' ||
      (claims.mode === 'voice_control' && claims.capabilities.includes('proposal:decide'))) &&
    claims.capabilities.includes('integration:write')
  );
}
